SOLR-16265 reduce memory usage of ContentWriter based requests in Http2SolrClient

stillalex opened the pull request:
https://issues.apache.org/jira/browse/SOLR-16265

<!--
_(If you are a project committer then you may remove some/all of the following template.)_

Before creating a pull request, please file an issue in the ASF Jira system for Solr:

* https://issues.apache.org/jira/projects/SOLR

For something minor (i.e. that wouldn't be worth putting in release notes), you can skip JIRA. 
To create a Jira issue, you will need to create an account there first.

The title of the PR should reference the Jira issue number in the form:

* SOLR-####: <short description of problem or changes>

SOLR must be fully capitalized. A short description helps people scanning pull requests for items they can work on.

Properly referencing the issue in the title ensures that Jira is correctly updated with code review comments and commits. -->


# Description

Initially started with addressing the TODO to remove extra buffer from the code (https://github.com/apache/solr/pull/1789/files#diff-db379c74efc9ff28acfd7571b07e16dbcdaa510783f2bee835b4a288927b5b3dR694). Ended up with removing all deprecated code to refresh implementation and bring it up to latest guidance.
It looks a bit more involved because some of the writers now need to push content after being setup, opposed to before style that was upfront setup of content.

To evaluate some performance aspect I ran exiting indexing benchmark. Numbers were equivalent, but heap footprint on benchmark client side saw a drop. posted screenshots below. if there are other tests I could run please let me know.


# Solution

Please provide a short description of the approach taken to implement your solution.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

stillalex commented:
Attaching memory footprint of client side during indexing benchmark. While perf numbers are very similar, memory footprint definitely sees a drop.

* main branch

<img width="1449" alt="bench-http2client-main" src="https://github.com/apache/solr/assets/61026/58c5d718-94a4-454a-ac09-470d5fedacc5">

* proposed patch

<img width="1449" alt="bench-http2client-patch" src="https://github.com/apache/solr/assets/61026/ef9e1547-4cb7-467c-ad1c-f0e4255eee9c">

risdenk commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -755,11 +762,12 @@ private Request fillContentStream(
         }
       }
-      req.content(content);
+      content.close();
```
Do we have to worry about closing this if there is an exception? not sure if `MultiPartRequestContent` can do try w/ resources?

stillalex replied:
good idea! seems it does so I will give it a go.

stillalex replied:
change is in!

epugh commented:
I wish we had more confidence in our tests (benchmark and unit) that a change like this that passed the tests could be assumed to correct ;-).   I like anything that removes deprecated code!

stillalex commented:
> I wish we had more confidence in our tests (benchmark and unit) that a change like this that passed the tests could be assumed to correct ;-). I like anything that removes deprecated code!

oh yeah, I completely agree. how can we increase that confidence? I feel like there are quite a few tests already (running 'gradlew check' is already a lengthy affair)

stillalex commented:
thank you @risdenk and @markrmiller for the reviews. I have added one more change which was a bit awkward. I needed the request url for the exception message inside the async listener, so I moved some methods around to build the url first. it's the least bad solution I could come up with. this doesn't change much of the PR, so I will go ahead and try to push this in once the builds complete.

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -438,81 +436,78 @@ public void send(OutStream outStream, SolrRequest<?> req, String collection) thr
+            @Override
+            public void onHeaders(Response response) {
+              super.onHeaders(response);
+              InputStreamResponseListener listener = this;
```
must this be defined a second time?

stillalex replied:
does not seem to be needed. weird, removed it now

dsmiley commented on solr/solrj/src/java/org/apache/solr/client/solrj/impl/Http2SolrClient.java:
```diff
@@ -659,62 +633,97 @@ private Request createRequest(SolrRequest<?> solrRequest, String collection)
-            // We're throwing this BAOS away, so no need to copy the byte[], just use the raw buf
-            new ByteBufferContentProvider(
-                contentWriter.getContentType(), ByteBuffer.wrap(baos.getbuf(), 0, baos.size())));
+        OutputStreamRequestContent content =
```
Just want to mention that cases like this are perfect for "var".  For next time...

stillalex replied:
agreed. I added a few of those below, this one got away. replaced it with latest commit

dsmiley reviewed (approved):
Love the perf analysis!

stillalex commented:
will push changes.txt to trigger a new build (I needed to do that anyway)
